### PR TITLE
Preliminary test of the octave-doctest library

### PR DESCRIPTION
**Pull request by denisrosset**
Added `octave-doctest` as a library (Git submodule).
Added a `replab_rundoctests` helper function, which runs coverage testing.

To test this branch, one needs to only do `replab_addpaths` and then `replab_rundoctests`.

Note: the use of `octave-doctest` uncovered problems in the class hierarchy (dead code), that I'll be fixing in master. Until then, the doctests won't run on Octave, only on Matlab (it will complain about missing base classes in Octave).

### CODE REVIEW

**Comment by denisrosset** [comment]
@jdbancal Updated branch and instructions.

**Comment by denisrosset** [comment]
> Yes, this is sad. The problem with removing this condition to satisfy octave though, is that we won't remember that this class should be inferior... and this can cause serious inconsistencies such as the appearance of non-commutative algebraic behaviors...

What's the problem and is it possible to work around it in user code? Would using old style classes help in that respect with Octave?

**Comment by jdbancal** [comment]
The problem is the following: if you have two classes `A` and `B`, and are unable to dictate that `A` comes before `B`, then class precedence goes to the first object encountered. Therefore:

 - when you call `A+B = plus(A,B)`, you are sent to `A.plus(B)`, which is what you want
 - when you call `B+A = plus(B,A)`, you are sent to `B.plus(A)`, which is not what you want

Typically, this would cause an error if you call `B+A` and class `B` doesn't know how to handle objects of type `B`. But still, in other words, the addition is effectively not commutative.

It is of course possible to counter this behavior, by redirecting to `A.plus(B)` inside `B.plus` if the argument is of type `A`, but this requires changing the code of class `B`. In your case, this requires changing the matlab class `sym` ;-P No need to change class `double`, though, this class is [always inferior to user classes](https://www.mathworks.com/help/matlab/matlab_oop/class-precedence.html#brf5okb-2).

Going back to old style classes... will bring new levels of problems which we don't have now. For instance [this bug](https://savannah.gnu.org/bugs/?54941) makes it essentially impossible to have old style classes inside a package in octave. I'm not sure what remain from our library if we have to avoid that, lol. I encountered this bug with CommutantVar, and it is the reason why I had to put all the methods of this class in one big ugly file :(

At minimum, I think it would be good to keep in the code a remark that says that ideally we would like to state that this class is superior to `sym`, otherwise we might loose this knowledge.

**Comment by denisrosset** [comment]
Thanks for the hat tip to old style classes. Let's not go there :)

Ok, so this is not so bad, if `double` is always inferior. `sym` is its own thing, and usually quite slow beyond simple examples if you go back and forth between Matlab and the symbolic engine. And Octave support of `sym` requires Python and `sympy`, which is far from the simplicity that RepLAB advocates!

All in all, this means that `RepLAB` will not compose well with other libraries... but in general, libraries do not work well together in the Matlab ecosystem, except when care is taken to ensure compatibility (example: QETLAB support for CVX).

The only thing on my mind is compatibility with Yalmip's `sdpvar`. How is the situation in that respect?

**Comment by jdbancal** [comment]
The place where I imagine we could get mixed with `sdpvar`s is currently in the `CommutantVar` class, because this class is at the interface, so supposed to deal with both objects. This is why I wanted to define `sdpvar` as inferior to `CommutantVar`, but had to retract from doing it because I'm not allowed to do this in octave... so it is currently not enforced. As a result, the current status is that when applying an operation on both a `CommutantVar` object and a `sdpvar` object, the `CommutantVar` object should come first...

Concretely, this means that is `M` is a `CommutantVar` matrix and `T` a `sdpvar` matrix, we should write
```
M <= T
```
instead of
```
T >= M
```
The last command will be handed onto yalmip, which is not aware of RepLAB, so will certainly not do what it should.

In the worst case, we could ask to modify yalmip's code, since it's open source. But I'ld not like to pollute that code. Especially when there is a matlab function meant to do the job on our side of the code! At some point the problem has to be solved at the root...

**Comment by denisrosset** [comment]
To alleviate the issue for end users, we should make it very convenient to declare a `EquivariantVar` which has no symmetry info and simply wraps a `sdpvar`. Then we tell users to not mix Yalmip base sdpvars and our vars.

We can revisit that later. The YALMIP interface in RepLAB does not need to be as polished as the rest of the code.

(And it's very convenient to be able to use Octave when teaching, so I like that we have this goal in mind!)

**Comment by jdbancal** [comment]
Following on https://github.com/replab/replab/pull/90#issuecomment-534314208 : actually, since yalmip is in the path when using CommutantVar objects, we should be able to declare sdpvar objects as inferior to CommutantVar ones, thus allowing for correct precedence.

However, for some reason, it seems that octave has yet another bug with this, it replies 
```
error: class not found: sdpvar
```
although the sdpvar class is obviously in the path:
```
octave:4> sdpvar(2)
Linear matrix variable 2x2 (symmetric, real, 3 variables)
Coeffiecient range: 1 to 1
octave:5> class(sdpvar(2))
ans = sdpvar
```
so for now correct precedence is left to the user

**Comment by jdbancal** [comment]
Coming back to the main subject of doctests:

Following our email discussion, it seems that the main role of doctests would be to:

 - Include examples that are run automatically in the documentation section of functions

A few points of interest:

1. This would replace the content of the current `Example` section of the documentation (actually, they should probably be put under this section). However, since it would be run automatically, the documentation would be kept up to date (including on a potential Sphinx API doc). I think this would be an appreciable feature: current examples in the documentation are not tested, hence they might not always be noticed when becoming outdated.

2. Doctests could be used to perform code coverage. The advantage compared to the current test folder framework being that:
    - Tests would be written next to the function they test, hence easing maintenance
    - Coverage increase is foreseen by using many small tests, which could be written in doctests

While I agree with point 1, and see a clear advantage there, I am not fully convinced about point 2. We have had some trouble already with making coverage tests quick, and included some optimization (such as use of global variables computed just once) which don't seem to have analogs in the doctest framework. So I'm pretty sure that if we only rewrite the current tests as doctests, we will certainly overcome the 50mn boundary of travis. Also, following this idea, every test needed for coverage purpose would have to appear in the documentation section of the function... seems like this might generate quite some text with low information content for the user. So I don't think we should rely on doctests for test coverage.

Finally, I see a third point worth mentioning:

3. We haven't tested much how the output integrates with the user documentation. In particular, it seems to include the full matlab output including `>>` symbols and names of variables such as `ans` in the following example:
```matlab
function s = S(n)
% Returns the symmetric group acting on n elements
%
% Alias for replab.Permutations(n)
%
% >> replab.S(3).order
% ans =
%     6
    s = replab.Permutations(n);
end
```
This might not really be an issue, but it would be good to experiment this behaviour further before deciding whether we should go full on or not.

So in short I think it would be good to experiment a bit more on it, considering using doctests for maintained examples, but not count on it for coverage.

**Comment by denisrosset** [comment]
I agree in general with your comments. To be precise, I was counting of doctests complementing the code coverage, especially for small helper methods where having a dedicated test smells like boilerplate (i.e. the `irreducible.nComponents` method, etc.). Going over the 50m limit is a good point, we'd need to have a flag for doctests as well, etc..., which may be a bit premature.

Let's agree to implement a few examples, see how they display, including in Sphinx and the command line, but let's not include them in the coverage tests.

**Comment by denisrosset** [comment]
Just a remark: this PR has a lot of conflicts because Octave, in combination with doctests, did not like the name of the `Settings` class.

I may replace `Settings` with `Parameters` and leave the rest of the code intact on `master`, just to avoid constant upgrading of this PR.

**Comment by denisrosset** [comment]
Octave-doctest is super fickle when used with new style classes. Code generation is the way to go, closing this.
